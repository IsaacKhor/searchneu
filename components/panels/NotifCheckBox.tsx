/*
 * This file is part of Search NEU and is licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import { uniqueId } from 'lodash';
import React, { ReactElement, useEffect, useState } from 'react';
import { Icon } from 'semantic-ui-react';
import IconCheckMark from '../icons/IconCheckmark';
import Keys from '../Keys';
import macros from '../macros';
import { Section } from '../types';
import user from '../user';

type NotifCheckBoxProps = {
  section: Section;
};

export default function NotifCheckBox({
  section,
}: NotifCheckBoxProps): ReactElement {
  const [checked, setChecked] = useState(
    user.isWatchingSection(Keys.getSectionHash(section))
  );
  const [notifSwitchId] = useState(uniqueId('notifSwitch-'));

  function onCheckboxClick(): void {
    if (checked) {
      user.removeSection(section);
      setChecked(false);
    } else {
      user.addSection(section);
      setChecked(true);
    }
  }

  useEffect(() => {
    function onUserUpdate(): void {
      setChecked(user.isWatchingSection(Keys.getSectionHash(section)));
    }
    user.registerUserChangeHandler(onUserUpdate);
    return () => user.unregisterUserChangeHandler(onUserUpdate);
  }, [section]);

  if (section.seatsRemaining > 5) {
    return (
      <div
        style={{ color: '#d3d3d3' }}
        data-tip="There are still seats remaining for this section"
        className="inlineBlock"
      >
        <Icon name="info circle" className="myIcon" />
      </div>
    );
  }

  return (
    <div
      data-tip="Sign up for notifications for this section"
      className="inlineBlock"
    >
      {macros.isMobile ? (
        <div
          className={
            checked ? 'notifSubscribeButton--checked' : 'notifSubscribeButton'
          }
          role="button"
          tabIndex={0}
          onClick={onCheckboxClick}
        >
          {checked && <IconCheckMark />}
          <span>{checked ? 'Subscribed' : 'Subscribe'}</span>
        </div>
      ) : (
        <div className="notifSwitch">
          <input
            checked={checked}
            onChange={onCheckboxClick}
            className="notif-switch-checkbox"
            id={notifSwitchId}
            type="checkbox"
          />
          <label className="notif-switch-label" htmlFor={notifSwitchId}>
            <span className="notif-switch-button" />
          </label>
        </div>
      )}
    </div>
  );
}
