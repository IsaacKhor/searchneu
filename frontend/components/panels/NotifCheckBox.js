/*
 * This file is part of Search NEU and is licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import React from 'react';
import { Checkbox, Icon } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import user from '../user';
import Keys from '../../../common/Keys';
import macros from '../macros';

// This file renders the checkboxes that control which sections a user signs up for
// notifications.
export default class NotifCheckBox extends React.Component {
  // the only thing required for you to pass in is what section this NotifCheckBox
  // was rendered in. We can get all the other data from that.
  static propTypes = {
    section: PropTypes.object.isRequired,
  }

  // creates a new Notification Check Box that has the section's
  // hash in section, and sets checked to be the initial value of if the section
  // is currently in the user or not
  constructor(props) {
    super(props);

    this.state = {
      // has a user already signed up for notifications on a section?
      checked: user.hasSectionAlready(Keys.getSectionHash(this.props.section)),

      // the section for a class that this NotifCheckBox is for
      section: this.props.section,
    };


    this.doChange = this.doChange.bind(this);
  }

    async componentDidMount() {


	  

    }

  // if the state is currently checked, uncheck, remove the section from the user's data
  // do opposite.
  // send data to backend
    async doChange() {
	if (!user.user) {
	    await user.downloadUserData();
	}
    if (this.state.checked) {
      user.removeSection(this.state.section);
      this.setState({ checked: false });
    } else {
      user.enrollSection(this.state.section);
      this.setState({ checked: true });
    }
    user.sendUserData();
  }


  // renders the proper checkbox. If there are still seats, then make it read
  // only, otherwise, set up callback on onChange
    render() {


    // one last check to ensure the state is correct... since sometimes the user isn't
    // in place by the time rendering has started.
    if (this.state.checked !== user.hasSectionAlready(Keys.getSectionHash(this.state.section))) {
      this.state.checked = !this.state.checked;
    }

    // if we have a section, and the section has seats remaining, it doesn't make
    // sense to sig up a user for notifications, so make the Checkbox readonly
    if (this.state.section && this.state.section.seatsRemaining) {
	return <Icon size='large' name='info circle' color='grey'/>;
    }

    // otherwise, return a checkbox that has the correct state
    return <Checkbox toggle checked={ this.state.checked } onChange={ this.doChange } />;
  }
}
