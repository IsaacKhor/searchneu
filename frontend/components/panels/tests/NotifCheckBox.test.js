/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import mockData from './mockData';
import NotifCheckBox from '../NotifCheckBox';

Enzyme.configure({ adapter: new Adapter() });

it('render the default notification checkbox', () => {
  const result = shallow(<NotifCheckBox section={ mockData.cs1210.sections[0] } />);
  expect(result).toMatchSnapshot();
});

it('render the read only notification checkbox', () => {
  const result = shallow(<NotifCheckBox section={ mockData.cs1210.sections[2] } />);
  expect(result).toMatchSnapshot();
});
