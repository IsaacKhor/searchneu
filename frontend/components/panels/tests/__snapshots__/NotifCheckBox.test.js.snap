// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the default notification checkbox 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NotifCheckBox
    section={
      Section {
        "classId": "1210",
        "crn": "13502",
        "host": "neu.edu",
        "lastUpdateTime": 1511131914029,
        "meetings": Array [
          Meeting {
            "endDate": "2017-12-07T00:00:00.000Z",
            "profs": Array [
              "Jennifer Anne Shire",
            ],
            "startDate": "2017-09-07T00:00:00.000Z",
            "times": Array [
              Array [
                Object {
                  "end": "1970-01-06T11:30:00.000Z",
                  "start": "1970-01-06T09:50:00.000Z",
                },
              ],
            ],
            "where": "Snell Engineering Center 108",
          },
        ],
        "online": false,
        "seatsCapacity": 19,
        "seatsRemaining": 0,
        "subject": "CS",
        "termId": "201810",
        "url": "https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201810&crn_in=13502",
        "waitCapacity": 0,
        "waitRemaining": 0,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "checked": false,
      "onChange": [Function],
      "toggle": true,
      "type": "checkbox",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "checked": false,
        "onChange": [Function],
        "toggle": true,
        "type": "checkbox",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`render the read only notification checkbox 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NotifCheckBox
    section={
      Section {
        "classId": "1210",
        "crn": "14404",
        "host": "neu.edu",
        "lastUpdateTime": 1511131914065,
        "meetings": Array [
          Meeting {
            "endDate": "2017-12-07T00:00:00.000Z",
            "profs": Array [
              "Jennifer Anne Shire",
            ],
            "startDate": "2017-09-07T00:00:00.000Z",
            "times": Array [
              Array [
                Object {
                  "end": "1970-01-06T13:25:00.000Z",
                  "start": "1970-01-06T11:45:00.000Z",
                },
              ],
            ],
            "where": "Ryder Hall 431",
          },
        ],
        "online": false,
        "seatsCapacity": 19,
        "seatsRemaining": 0,
        "subject": "CS",
        "termId": "201810",
        "url": "https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201810&crn_in=14404",
        "waitCapacity": 0,
        "waitRemaining": 0,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "checked": false,
      "onChange": [Function],
      "toggle": true,
      "type": "checkbox",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "checked": false,
        "onChange": [Function],
        "toggle": true,
        "type": "checkbox",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
